clippy_lints/src/methods/mod.rs

```diff
@@ -3284,6 +3285,35 @@ declare_clippy_lint! {
     "calling `.drain(..).collect()` to move all elements into a new collection"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `Iterator::fold` with a type that implements `Try`.
+    ///
+    /// ### Why is this bad?
+    /// This should use `try_fold` instead, which short-circuits on failure, thus opening the door
+    /// for additional optimizations not possible with `fold` as rustc can guarantee the function is
+    /// never called on `None`, `Err`, etc., alleviating otherwise necessary checks. It's also
+    /// slightly more idiomatic.
+    ///
+    /// ### Known issues
+    /// This lint doesn't take into account whether a function does something on the failure case,
+    /// i.e., whether short-circuiting will affect behavior. Refactoring to `try_fold` is not
+    /// desirable in those cases.
+    ///
+    /// ### Example
+    /// ```rust
+    /// vec![1, 2, 3].iter().fold(Some(0i32), |sum, i| sum?.checked_add(*i));
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// vec![1, 2, 3].iter().try_fold(0i32, |sum, i| sum.checked_add(*i));
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub MANUAL_TRY_FOLD,
+    perf,
```

Comment:
The "known problem" isn't really an issue, imo. The known problem is referring to something like
```rust
x.iter().fold(Some(1), |mut acc, i| acc?.checked_add(i));
```
(pseudocode)

This will continue returning `None` if the last one was `None`, as `acc?` will return the `None`.

There are very few cases where that's is the desired behavior, most of the time it'd continue returning `None` if it's passed `None` due to usage of `?` or `and_then`, unless the programmer explicitly handles the `None` case, doing something with it, it's just a performance gain

Reply:
I also agree that, if this behaviour is well-documented (as I think it is, as it appears in the `std` documentation and in the lint's documentation) it doesn't really represent an issue.

Also, I don't think I've ever seen a `fold` that handles a `None` case, but I'll make a Github regex search to confirm how many times does this happen

Reply:
Yeah, I'm just a more cautious when it comes to warn-by-default lints. I would really like nightly lints at some point, but first I finally want to stabilize lint reasons.

I believe it should be fine :)

Reply:
@Centri3 

> Can be changed to try_fold just fine, as it does nothing in the None case.

> There are very few cases where that's the desired behavior, as it doesn't change the outcome at all. You can short-circuit there and it changes nothing. Unless the programmer explicitly handles the None case, doing something with it, it's just a performance gain

In that specific case, sure.  But sometimes `fold` is used for things that are fallible, but you don't want it to short-circuit.

Example:

```Rust
            self.public_key
                .public_subkeys
                .iter()
                .filter(|sub_key| {
                    if sub_key.key_id().as_ref() == key_id.as_ref() {
                        log::trace!(
                            "Found a matching key id {:?} == {:?}",
                            sub_key.key_id(),
                            key_id
                        );
                        true
                    } else {
                        log::trace!("Not the one we want: {:?}", sub_key);
                        false
                    }
                })
                .fold(
                    Err(Error::KeyNotFoundError {
                        key_ref: format!("{:?}", key_id),
                    }),
                    |previous_res, sub_key| {
                        if previous_res.is_err() {
                            log::trace!("Test next candidate subkey");
                            signature.verify(sub_key, &mut data).map_err(|e| {
                                Error::VerificationError {
                                    source: Box::new(e),
                                    key_ref: format!("{:?}", sub_key.key_id()),
                                }
                            })
                        } else {
                            log::trace!("Signature already verified, nop");
                            Ok(())
                        }
                    },
                )
```

This code initializes with an error of "KeyNotFoundError", and replaces it with an error of "VerificationError" after a subkey has been tried, but still tries all of the subkeys before doing so.  If there's a successful result it short-circuits but in the `Ok` case rather than the error case.

This might be rare enough to not be worth worrying about - I'm not suggesting that the lint should be reverted necessarily.  Just pointing out a valid case where as of the moment I'm going to need to mark the lint ignored or rewrite the code.